cpp/src/io/parquet/reader_impl_chunking.cu

```diff
@@ -752,83 +754,92 @@ std::vector<row_range> compute_page_splits_by_row(device_span<cumulative_page_in
  *
- * This function handles the case where `pages` is only a subset of all available
- * pages in `chunks`.
+ * To avoid multiple calls to the decompression kernel, we batch pages by codec type, where the
+ * batch can include both dictionary and non-dictionary pages. This allows us to decompress all
+ * pages of a given codec type in one go.
  *
  * @param chunks List of column chunk descriptors
- * @param pages List of page information
- * @param dict_pages If true, decompress dictionary pages only. Otherwise decompress non-dictionary
- * pages only.
+ * @param pass_pages List of page information for the pass
+ * @param subpass_pages List of page information for the subpass
  * @param stream CUDA stream used for device memory operations and kernel launches
  *
  * @return Device buffer to decompressed page data
  */
-[[nodiscard]] rmm::device_buffer decompress_page_data(
-  cudf::detail::hostdevice_span<ColumnChunkDesc const> chunks,
-  cudf::detail::hostdevice_span<PageInfo> pages,
-  bool dict_pages,
+[[nodiscard]] std::pair<rmm::device_buffer, rmm::device_buffer> decompress_page_data(
+  host_span<ColumnChunkDesc const> chunks,
+  host_span<PageInfo> pass_pages,
+  host_span<PageInfo> subpass_pages,
   rmm::cuda_stream_view stream)
 {
   CUDF_FUNC_RANGE();
 
-  auto for_each_codec_page = [&](Compression codec, std::function<void(size_t)> const& f) {
-    for (size_t p = 0; p < pages.size(); p++) {
-      if (chunks[pages[p].chunk_idx].codec == codec &&
-          ((dict_pages && (pages[p].flags & PAGEINFO_FLAGS_DICTIONARY)) ||
-           (!dict_pages && !(pages[p].flags & PAGEINFO_FLAGS_DICTIONARY)))) {
-        f(p);
-      }
-    }
-  };
-
-  // Count the exact number of compressed pages
-  size_t num_comp_pages    = 0;
-  size_t total_decomp_size = 0;
-
-  struct codec_stats {
-    Compression compression_type  = UNCOMPRESSED;
-    size_t num_pages              = 0;
-    int32_t max_decompressed_size = 0;
-    size_t total_decomp_size      = 0;
-  };
-
-  std::array codecs{codec_stats{GZIP},
+  std::array codecs{codec_stats{BROTLI},
+                    codec_stats{GZIP},
+                    codec_stats{LZ4_RAW},
                     codec_stats{SNAPPY},
-                    codec_stats{BROTLI},
-                    codec_stats{ZSTD},
-                    codec_stats{LZ4_RAW}};
+                    codec_stats{ZSTD}};
 
   auto is_codec_supported = [&codecs](int8_t codec) {
     if (codec == UNCOMPRESSED) return true;
     return std::find_if(codecs.begin(), codecs.end(), [codec](auto& cstats) {
              return codec == cstats.compression_type;
            }) != codecs.end();
   };
-  CUDF_EXPECTS(std::all_of(chunks.host_begin(),
-                           chunks.host_end(),
-                           [&is_codec_supported](auto const& chunk) {
-                             return is_codec_supported(chunk.codec);
-                           }),
-               "Unsupported compression type");
 
+  for (auto& chunk : chunks) {
```

Comment:
nit:

```suggestion
  for (auto const& chunk : chunks) {
```

Reply:
it's already const becauce the parameter is `host_span<ColumnChunkDesc const> chunks`, but it doesn't hurt to be explicit

Reply:
Oh yeah, duh. I missed that.